config/prow/cluster/plank_rbac.yaml

```diff
@@ -22,10 +22,9 @@ rules:
       - update
```

Comment:
This is a pretty incompatible change and will break all existing deployments. In order to actually leverage this you need to at least update `prow/cmd/plank/main.go` to not watch one but all namespaces. I believe sinker will need updating as well.

And we have to think about some kind of strategy that does not involve breaking existing deployments. We could for example in the main test if we can do a `List` with `MaxResult: 1` in a namesapce that is guaranteed to exist but not the `ProwJob` namespace, like for example `kube-system`. If we get a 403 we only use the `ProwJob` namespace and log a warning, otherwise we use all namespaces.

The unfortunate side effect is that if someone creates a job with a different namespace but uses the namespace-scoped rbac, it will silently not work because `plank` doesn't see the job.

Reply:
Everything is better with flags?

- We could expose a flag or config parameter that would allow us to set the selector namespace... Defaulting to podnamespace?
- Or we could have a flag or config parameter called something like "allow_namespace_override"? This new config option would then require the cluster rbac.

How does that sound?

Reply:
tbh if possible I would much prefer if we not added even more configuration options :grimacing: We have way too many already

Reply:
Two things on this 😸 

Start-up probes in kubernetes land are normally used for things that `don't change at runtime` and that are normally `environmental capabilities` (like cloud-provider capabilities or kubernetes api server api availability). RBAC in kubernetes is detached from runtime items, so we would be attaching a dynamic non-runtime item to a runtime probing technique (not ideal and would probably cause confusion). Also, we would need to restart services to change RBAC (not normally done in kubernetes land, since RBAC is decoupled from runtime).

I think if stable eco-system patterns exist for certain approaches, we should use those, that's what people are expecting... anything else will probably confuse people in the beginning and frustrate them later (when they find out you deviated from normal practices).

Thinking through this though... we can "have both". What if we leave the `namespaced` role for `podNamespace` and add another clusterRole? That should allow a more gradual change workflow? We would still need someway of changing the selector namespace (I would go with flag or config param).

Reply:
We can not have both because we use a watch and we can do that either for all namespaces or for one namespace.

I am mainly opposed to introduce yet another config switch. Either we decide that giving perms for all namespaces is the right thing or we don't.

The probing for perms is done to be able to give Prow users some time to migrate and warn them that they need to do something. Compatibility is an annoying thing that does not always allow you to follow "best practices" but its also extremely important if you are building software that is being used.

Reply:
> I am mainly opposed to introduce yet another config switch. Either we decide that giving perms for all namespaces is the right thing or we don't.

I know k8s prow currently uses a `trusted` k8s cluster for trusted workloads, do you guys have any desire to shard trust by namespace? Or would you prefer to create a whole k8s cluster for each trust shard?

Do we need a KEP for this? Should we chat more in slack? How do we move forward on this?

Reply:
If we use a watch for pods, we need to update this manifest... it only shows `list` permissions.

Reply:
Right, we don't do it right now in plank v1 but we do it in plank v2 and the former will eventually get removed (and we should not add features to one but not the others).

Yeah I am not completely sure about the best approach. We could also make the pod namespace a part of the build cluster (with default to the global default pod namespace) and then you could have a virtual cluster `mybuildcluster-other-namespace`.

And for the record, this is not something I can just decide/approve alone. Maybe create an issue first and ping erick, steve and me and bring it up in the biweekly-meeting?

Reply:
> Right, we don't do it right now in plank v1 but we do it in plank v2 and the former will eventually get removed (and we should not add features to one but not the others).

I wasn't sure how far along plank v2 had come... should we possibly start with that one instead? It was much easier to make this change to v1, I had strange errors in v2 (I assume it's due to the informer pattern).

> Yeah I am not completely sure about the best approach. We could also make the pod namespace a part of the build cluster (with default to the global default pod namespace) and then you could have a virtual cluster mybuildcluster-other-namespace.

We spoke about this too... this was originally how it seemed something like this should be done, but I ran into the hardcoded `podNamespace` and then saw the prowJob namespace parameter, and then we arrived here :D You guys let me know what you think would be the best starting point and direction.

> And for the record, this is not something I can just decide/approve alone. Maybe create an issue first and ping erick, steve and me and bring it up in the biweekly-meeting?

No problem... we already have an issue for this (mentioned via `fixes` in the PR comment), I'll ping steve and fejta on it.

Thank you for the help, it is much appreciated!

Reply:
> I wasn't sure how far along plank v2 had come... should we possibly start with that one instead? 

Yeah I think that would be better. It definitely isn't totally broken and supposed to replace plank entirely in the future, we replaced Plank with it on june fifth for openshift. If you have problems with it, please open an issue :)